File: VC_HIL/python/inverterhil/evaluate_can_diagnostics.py
import math


def _flag(value, name):
    if not isinstance(value, (bool, int, float)) or not math.isfinite(float(value)) or float(value) not in (0, 1):
        raise ValueError(f"{name} must be logical 0 or 1.")
    return bool(value)


def evaluateCanDiagnostics(writeSucceeded, backlog, receiveOverrun, errorWarning, busOff):
    if not isinstance(writeSucceeded, (list, tuple)) or len(writeSucceeded) != 9 or any(not isinstance(x, bool) for x in writeSucceeded):
        raise ValueError("writeSucceeded must be a 1x9 logical vector.")
    flags = [_flag(v, n) for v, n in zip((backlog, receiveOverrun, errorWarning, busOff),
                                         ("backlog", "receiveOverrun", "errorWarning", "busOff"))]
    mask = sum(1 << i for i, ok in enumerate(writeSucceeded) if not ok)
    return {"writeSucceeded": list(writeSucceeded), "failedWriteMask": mask, "backlog": flags[0],
            "receiveOverrun": flags[1], "errorWarning": flags[2], "busOff": flags[3],
            "cycleSucceeded": all(writeSucceeded) and not any(flags)}

File: VC_HIL/inverter/evaluateCanDiagnostics.m
function result = evaluateCanDiagnostics(writeSucceeded, backlog, ...
        receiveOverrun, errorWarning, busOff)
%EVALUATECANDIAGNOSTICS Decide whether a nine-frame cycle really succeeded.

if ~islogical(writeSucceeded) || ~isequal(size(writeSucceeded), [1 9])
    error('inverterhil:MalformedDiagnostics', ...
        'writeSucceeded must be a 1x9 logical vector.');
end
backlog = logicalScalar(backlog, 'backlog');
receiveOverrun = logicalScalar(receiveOverrun, 'receiveOverrun');
errorWarning = logicalScalar(errorWarning, 'errorWarning');
busOff = logicalScalar(busOff, 'busOff');

result.writeSucceeded = writeSucceeded;
result.failedWriteMask = uint16(0);
for index = 1:9
    if ~writeSucceeded(index)
        result.failedWriteMask = bitor(result.failedWriteMask, ...
            bitshift(uint16(1), index - 1));
    end
end
result.backlog = backlog;
result.receiveOverrun = receiveOverrun;
result.errorWarning = errorWarning;
result.busOff = busOff;
result.cycleSucceeded = all(writeSucceeded) && ~backlog && ...
    ~receiveOverrun && ~errorWarning && ~busOff;
end

function value = logicalScalar(value, name)
if ~(islogical(value) || isnumeric(value)) || ~isscalar(value) || ...
        ~isfinite(double(value)) || ~(double(value) == 0 || double(value) == 1)
    error('inverterhil:MalformedDiagnostics', ...
        '%s must be logical 0 or 1.', name);
end
value = logical(value);
end
